explorer/server.go

```diff
@@ -0,0 +1,218 @@
+}
+
+func (a *App) listNodes(w http.ResponseWriter, r *http.Request) {
+
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		w.Write([]byte("500 - Something bad happened!"))
+		return
+	}
+	log.Debug().Str("request_body", string(body)).Msg("request from external agent")
+
+	farmID := r.URL.Query().Get("farm_id")
+	isSpecificFarm := ""
+	if farmID != "" {
+		isSpecificFarm = fmt.Sprintf(",where:{farmId_eq:%s}", farmID)
+	} else {
+		isSpecificFarm = ""
+	}
+	page := r.URL.Query().Get("page")
+	if page == "" {
+		page = "0"
+	}
+	pageNumber, err := strconv.Atoi(page)
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		w.Write([]byte("400 - Bad page number!"))
+		return
+	}
+	offset := 0
+	if pageNumber > 1 {
+		offset = pageNumber * 50
+	}
+	log.Debug().Str("page number", fmt.Sprint(pageNumber)).Msg("Requested page number")
+	queryString := fmt.Sprintf(`
+	{
+		nodes(limit:50,offset:%d,%s){
+			version          
+			id
+			nodeId        
+			farmId          
+			twinId          
+			country
+			gridVersion  
+			city         
+			uptime           
+			created          
+			farmingPolicyId
+			cru
+			mru
+			sru
+			hru
+		publicConfig{
+			gw4
+			ipv4
+			ipv6
+			gw6
+		  }
+		}
+	}
+	`, offset, isSpecificFarm)
+
+	farmsData := query(queryString)
+
+	jsonBytes := []byte(farmsData)
+
+	w.Write(jsonBytes)
+}
+
+func (a *App) getNode(w http.ResponseWriter, r *http.Request) {
+	nodeID := mux.Vars(r)["node_id"]
+
+	value, err := a.GetRedisKey(fmt.Sprintf("GRID3NODE:%s", nodeID))
+
+	if err != nil {
+		log.Warn().Str("Couldn't find entry to redis", string(err.Error())).Msg("")
+
+	}
+	if value != "" {
+		w.Write([]byte(value))
+		return
+	}
+	TwinID := getNodeTwinID(nodeID)
+	if TwinID < 1 {
+		value, err := json.Marshal("Couldn't find node ID.")
+		if err != nil {
+			log.Error().Err(errors.Wrap(err, "Couldn't get node twin ID")).Msg("")
+			w.WriteHeader(http.StatusInternalServerError)
+			w.Write([]byte("500 - Something bad happened!"))
+			return
+		}
+		w.Write(value)
+		return
+	}
+
+	nodeClient := NewNodeClient(TwinID, a.rmb)
+	nodeCapacity, err := nodeClient.NodeStatistics(a.ctx)
+	if err != nil {
+		log.Error().Err(errors.Wrap(err, "Couldn't get node statistics")).Msg("connection error")
+		w.WriteHeader(http.StatusInternalServerError)
+		w.Write([]byte("500 - Something bad happened!"))
+		return
+	}
+
+	totalCapacity, err := json.Marshal(nodeCapacity)
+	if err != nil {
+		log.Error().Err(errors.Wrap(err, "Couldn't get node statistics")).Msg("connection error")
+		w.WriteHeader(http.StatusInternalServerError)
+		w.Write([]byte("500 - Something bad happened!"))
+		return
+	}
+
+	w.Write(totalCapacity)
+
+	// caching for 30 mins
+	err = a.SetRedisKey(fmt.Sprintf("GRID3NODE:%s", nodeID), totalCapacity, 30*60)
+	if err != nil {
+		log.Error().Err(errors.Wrap(err, "Couldn't cache to redis")).Msg("connection error")
+		w.WriteHeader(http.StatusInternalServerError)
```

Comment:
You aware that it's IMPOSSIBLE to change the header after sending the body ? because header comes first. So once u did a w.Write() (linke 171) u already implied a response code of 200 and this can't change. Hence the redis set must be done first before writing the response. 

I also believe that failure to set the key in your local cache should not send an error to the user. he doesn't care if your redis is down if u already have a good answer for him. too bad your code will still have to make another rmb call to get the info next time he asks for the same data.